### ISSUE TITLE
Failure to single let-bound subexpression with a type signature

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#296]
Originally observed in https://github.com/goldfirere/singletons/pull/295#discussion_r164288565. This program:

```haskell
import Data.Kind

data MyProxy (a :: Type) = MyProxy

f :: forall a. MyProxy a -> MyProxy a
f MyProxy = 
  let x = let z :: MyProxy a
              z = MyProxy in z
  in x
```

Will correctly promote, but not correctly single, as this program demonstrates:

```haskell
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

import Data.Kind
import Data.Singletons.Prelude
import Data.Singletons.TH

$(singletons [d|
  data MyProxy (a :: Type) = MyProxy

  f :: forall a. MyProxy a -> MyProxy a
  f MyProxy =
    let x = let z :: MyProxy a
                z = MyProxy in z
    in x
  |])
```
```
$ /opt/ghc/8.4.1/bin/ghc Bug.hs -ddump-splices
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )
Bug.hs:(14,3)-(22,4): Splicing declarations
    singletons
      [d| f_a32l ::
            forall a_a3vI. MyProxy_a32j a_a3vI -> MyProxy_a32j a_a3vI
          f_a32l MyProxy_a32k
            = let
                x_a3vJ
                  = let
                      z_a3vK :: MyProxy_a32j a_a3vI
                      z_a3vK = MyProxy_a32k
                    in z_a3vK
              in x_a3vJ
          
          data MyProxy_a32j (a_a3vH :: Type) = MyProxy_a32k |]
  ======>
    data MyProxy_a6jb (a_a6jd :: Type) = MyProxy_a6jc
    f_a6ja :: forall a_a6j9. MyProxy_a6jb a_a6j9 -> MyProxy_a6jb a_a6j9
    f_a6ja MyProxy_a6jc
      = let
          x_a6je
            = let
                z_a6jf :: MyProxy_a6jb a_a6j9
                z_a6jf = MyProxy_a6jc
              in z_a6jf
        in x_a6je
    type MyProxySym0 = MyProxy_a6jc
    type Let6989586621679034055ZSym0 = Let6989586621679034055Z
    type family Let6989586621679034055Z :: MyProxy_a6jb a_a6j9 where
      Let6989586621679034055Z = MyProxySym0
    type Let6989586621679034054XSym0 = Let6989586621679034054X
    type family Let6989586621679034054X where
      Let6989586621679034054X = Let6989586621679034055ZSym0
    type FSym1 (t_a6jg :: MyProxy_a6jb a6989586621679034043) = F t_a6jg
    instance SuppressUnusedWarnings FSym0 where
      suppressUnusedWarnings
        = snd ((GHC.Tuple.(,) FSym0KindInference) GHC.Tuple.())
    data FSym0 (l_a6jh :: TyFun (MyProxy_a6jb a6989586621679034043) (MyProxy_a6jb a6989586621679034043))
      = forall arg_a6ji. SameKind (Apply FSym0 arg_a6ji) (FSym1 arg_a6ji) =>
        FSym0KindInference
    type instance Apply FSym0 l_a6jh = F l_a6jh
    type family F (a_a6jj :: MyProxy_a6jb a_a6j9) :: MyProxy_a6jb a_a6j9 where
      F MyProxy_a6jc = Let6989586621679034054XSym0
    sF ::
      forall (t_a6jm :: MyProxy_a6jb a_a6j9).
      Sing t_a6jm -> Sing (Apply FSym0 t_a6jm :: MyProxy_a6jb a_a6j9)
    sF SMyProxy
      = let
          sX :: Sing Let6989586621679034054XSym0
          sX
            = let
                sZ :: Sing (Let6989586621679034055ZSym0 :: MyProxy_a6jb a_a6j9)
                sZ = SMyProxy
              in sZ
        in sX
    data instance Sing (z_a6jn :: MyProxy_a6jb a_a6jd)
      where SMyProxy :: Sing MyProxy_a6jc
    type SMyProxy = (Sing :: MyProxy_a6jb a_a6jd -> Type)
    instance SingKind a_a6jd => SingKind (MyProxy_a6jb a_a6jd) where
      type Demote (MyProxy_a6jb a_a6jd) = MyProxy_a6jb (Demote a_a6jd)
      fromSing SMyProxy = MyProxy_a6jc
      toSing MyProxy_a6jc = SomeSing SMyProxy
    instance SingI MyProxy_a6jc where
      sing = SMyProxy

Bug.hs:14:3: error:
    • Could not deduce: a ~ a1
      from the context: t ~ 'MyProxy
        bound by a pattern with constructor:
                   SMyProxy :: forall a. Sing 'MyProxy,
                 in an equation for ‘sF’
        at Bug.hs:(14,3)-(22,4)
      ‘a’ is a rigid type variable bound by
        the type signature for:
          sF :: forall a (t :: MyProxy a). Sing t -> Sing (Apply FSym0 t)
        at Bug.hs:(14,3)-(22,4)
      ‘a1’ is a rigid type variable bound by
        the type signature for:
          sX :: forall a1. Sing Let6989586621679034054XSym0
        at Bug.hs:(14,3)-(22,4)
      Expected type: Sing Let6989586621679034054XSym0
        Actual type: Sing Let6989586621679034055ZSym0
    • In the expression: sZ
      In the expression:
        let
          sZ :: Sing (Let6989586621679034055ZSym0 :: MyProxy a_a6j9)
          sZ = SMyProxy
        in sZ
      In an equation for ‘sX’:
          sX
            = let
                sZ :: Sing (Let6989586621679034055ZSym0 :: MyProxy a_a6j9)
                sZ = SMyProxy
              in sZ
    • Relevant bindings include
        sZ :: Sing Let6989586621679034055ZSym0 (bound at Bug.hs:14:3)
        sX :: Sing Let6989586621679034054XSym0 (bound at Bug.hs:14:3)
        sF :: Sing t -> Sing (Apply FSym0 t) (bound at Bug.hs:14:3)
   |
14 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

From the `-ddump-splices` output, we can see why:

```haskell
    sF ::
      forall (t_a6jm :: MyProxy_a6jb a_a6j9).
      Sing t_a6jm -> Sing (Apply FSym0 t_a6jm :: MyProxy_a6jb a_a6j9)
    sF SMyProxy
      = let
          sX :: Sing Let6989586621679034054XSym0
          sX
            = let
                sZ :: Sing (Let6989586621679034055ZSym0 :: MyProxy_a6jb a_a6j9)
                sZ = SMyProxy
              in sZ
        in sX
```

GHC is unable to conclude that the `a1`—which is implicitly quantified in the type signature of `sX :: Sing Let6989586621679034054XSym0` —is the same `a` as in `sF`'s type signature.

I would bet that we could fix this by making that `a1` argument to `Let6989586621679034054XSym0` visible and propagating that change through to its defunctionalization symbols, but I don't know how much effort that would entail.

**Comment by goldfirere** [comment]
When singling the `let x = ...`, why do we add a type signature at all? I conjecture that we do this so that, if the definition is a function, the type signature can bring scoped type variables into scope. Back when I first implemented this, TH didn't support type annotations in patterns, so that wasn't an option. However, this is now fully supported (including in th-desugar).

So, I wonder if it wouldn't be better to omit the type signature on `let`s if the original code also omits the type signature, and use pattern annotations always to bind scoped type variables. That might sidestep this whole issue.

**Comment by RyanGlScott** [comment]
I'm not sure I follow. Are you saying that instead of generating this code:

```haskell
    sF ::
      forall (t_a6jm :: MyProxy_a6jb a_a6j9).
      Sing t_a6jm -> Sing (Apply FSym0 t_a6jm :: MyProxy_a6jb a_a6j9)
    sF SMyProxy
      = let
          sX :: Sing Let6989586621679034054XSym0
          sX
            = let
                sZ :: Sing (Let6989586621679034055ZSym0 :: MyProxy_a6jb a_a6j9)
                sZ = SMyProxy
              in sZ
        in sX
```

We should instead generate this?

```haskell
    sF ::
      forall (t_a6jm :: MyProxy_a6jb a_a6j9).
      Sing t_a6jm -> Sing (Apply FSym0 t_a6jm :: MyProxy_a6jb a_a6j9)
    sF SMyProxy
      = let
          sX
            = let
                sZ = SMyProxy
              in sZ
        in sX
```

In other words, just delete code from `singletons`? :)

**Comment by goldfirere** [comment]
Not quite, to both questions.

Note that `z` has a type signature originally, so it should keep a type signature after singling. This might be necessary to constraint the type/kind somehow.

And it's not just deleting code. For a let-bound function, `singletons` relies on the type signature to bring scoped type variables into scope, even if the original function declaration omits the type signature. So, if we omit signatures when the original code does, we'll need a new mechanism to grab the type variables, and that's why I've proposed pattern annotations.

**Comment by RyanGlScott** [comment]
OK, it seems that you're suggesting that we axe the type signature for `sX`, which I'm totally on board for. Your second suggestion confuses me—you suggest using pattern signatures, but where? `sZ` certainly doesn't use any patterns, so I have a hard time guessing at what you're hinting at.

**Comment by goldfirere** [comment]
I'm worried about singling, say

```
f x = ...
```

without a type signature. If we don't provide a type signature in the singled code, then the RHS might mention type variables that aren't in scope, because the explicit type signature is normally what brings such variables into scope. So we'd need a way to bring type variables into scope without a signature, which is by pattern annotations.

Does that help?

**Comment by RyanGlScott** [comment]
What happens if we single this definition?

```haskell
data MyProxy (a :: Type) = MyProxy

f MyProxy =
  let x = let z :: MyProxy a
              z = MyProxy in z
  in x
```

Interestingly enough, this actually typechecks when singled! (Perhaps due to a lucky fluke.) We currently generate:

```haskell
sF :: forall arg_ak2C. Sing arg_ak2C -> Sing (Apply FSym0 arg_ak2C)
sF SMyProxy
  = let
      sX :: Sing Let6989586621679086832XSym0
      sX
        = let
            sZ :: Sing (Let6989586621679086833ZSym0 :: MyProxy_ak2q a_ak2u)
            sZ = SMyProxy
          in sZ
    in sX
```

My question to you, is: what should change here? We could safely drop the `sX :: Sing Let6989586621679086832XSym0` type signature, but after that I'm unclear. You say you want to bring the type variable `a_ak2u` into scope via a pattern signature, but where exactly?

**Comment by goldfirere** [comment]
It should just drop the signature on `sX` and explicitly quantify the `a_ak2u` variable (in the signature for `sZ`). Indeed, I'm surprised the latter isn't already the case, based on your comments that singletons explicitly quantifies all the time, anyway.

The extra stuff to bind more scoped type variables matters only on *functions* with arguments. Each argument will correspond to a singletons-generated type variable that must be brought into scope.

**Comment by RyanGlScott** [comment]
> Indeed, I'm surprised the latter isn't already the case, based on your comments that singletons explicitly quantifies all the time, anyway.

That surprises me too! I thought my [explicitly bind kind variables patch](https://github.com/goldfirere/singletons/pull/298) would catch that, but it didn't due to an enormous oversight. This should be fixed as of https://github.com/goldfirere/singletons/pull/298/commits/370ccf39d76d6be79f133949bc939c211286c1b9.

> The extra stuff to bind more scoped type variables matters only on *functions* with arguments. Each argument will correspond to a singletons-generated type variable that must be brought into scope.

I apologize for being so slow, but I'm having the hardest time envisioning this. Let's say we have an example where we locally `let`-bind something with arguments, like so:

```haskell
  f :: forall a. a -> a
  f x =
    let g y = y
    in g x
```

Currently, this singles to:

```haskell
    sF ::
      forall a_aneu (t_aneO :: a_aneu).
      Sing t_aneO -> Sing (Apply FSym0 t_aneO :: a_aneu)
    sF (sX :: Sing x_aneD)
      = let
          sG ::
            forall arg_aneP.
            Sing arg_aneP
            -> Sing (Apply (Let6989586621679099112GSym1 x_aneD) arg_aneP)
          sG (sY :: Sing y_aneN) = sY
        in
          (applySing ((singFun1 @(Let6989586621679099112GSym1 x_aneD)) sG))
            sX
```

You are proposing to:

1. Remove the type signature for `sG`
2. Bind the types _ in the patterns _

Can you fill in the blanks?

**Comment by goldfirere** [comment]
Huh. Maybe I'm the one who's misinformed.

It seems we already use pattern annotations to bind tyvars, much to my surprise. So indeed I've been spouting nonsense. Sorry about that!

Yes, deleting the code in the `Nothing` branch of `singTySig` should work nicely, I think.

**Comment by RyanGlScott** [comment]
I tried your suggestion in https://github.com/goldfirere/singletons/issues/296#issuecomment-363273001, but alas, it seems nothing is ever easy. This plan falls over on `foldr`:

```haskell
    foldr                   :: (a -> b -> b) -> b -> [a] -> b
    foldr k z = go
              where
                go []     = z
                go (y:ys) = y `k` go ys
```

Before, this would single to:

```haskell
    sFoldr ::
      forall a_a1q2u
             b_a1q2v
             (t_a1q7x :: (Data.Singletons.Internal.~>) a_a1q2u ((Data.Singletons.Internal.~>) b_a1q2v b_a1q2v))
             (t_a1q7y :: b_a1q2v)
             (t_a1q7z :: [a_a1q2u]).
    sFoldr
      (sK :: Sing k_aiSK)
      (sZ :: Sing z_aiSL)
      (sA_6989586621679082376 :: Sing a_6989586621679082376_aiSM)
      = (Data.Singletons.Internal.applySing
           (let
              sGo ::
                forall arg_aiTS.
                Sing arg_aiTS
                -> Sing (Data.Singletons.Internal.Apply (Let6989586621679082381GoSym3 k_aiSK z_aiSL a_6989586621679082376_aiSM) arg_aiTS)
              sGo SNil = sZ
              sGo (SCons (sY :: Sing y_aiT7) (sYs :: Sing ys_aiT8))
                = (Data.Singletons.Internal.applySing
                     ((Data.Singletons.Internal.applySing sK) sY))
                    ((Data.Singletons.Internal.applySing
                        ((Data.Singletons.Internal.singFun1
                            @(Let6989586621679082381GoSym3 k_aiSK z_aiSL a_6989586621679082376_aiSM))
                           sGo))
                       sYs)
            in
              (Data.Singletons.Internal.singFun1
                 @(Let6989586621679082381GoSym3 k_aiSK z_aiSL a_6989586621679082376_aiSM))
                sGo))
          sA_6989586621679082376
```

If this is changed to leave off the type signature for `sGo`:

```haskell
    sFoldr ::
      forall a_a1q2u
             b_a1q2v
             (t_a1q7x :: (Data.Singletons.Internal.~>) a_a1q2u ((Data.Singletons.Internal.~>) b_a1q2v b_a1q2v))
             (t_a1q7y :: b_a1q2v)
             (t_a1q7z :: [a_a1q2u]).
    sFoldr
      (sK :: Sing k_afNP)
      (sZ :: Sing z_afNQ)
      (sA_6989586621679070539 :: Sing a_6989586621679070539_afNR)
      = (Data.Singletons.Internal.applySing
           (let
              sGo SNil = sZ
              sGo (SCons (sY :: Sing y_afOc) (sYs :: Sing ys_afOd))
                = (Data.Singletons.Internal.applySing
                     ((Data.Singletons.Internal.applySing sK) sY))
                    ((Data.Singletons.Internal.applySing
                        ((Data.Singletons.Internal.singFun1
                            @(Let6989586621679070544GoSym3 k_afNP z_afNQ a_6989586621679070539_afNR))
                           sGo))
                       sYs)
            in
              (Data.Singletons.Internal.singFun1
                 @(Let6989586621679070544GoSym3 k_afNP z_afNQ a_6989586621679070539_afNR))
                sGo))
          sA_6989586621679070539
```

Then GHC complains:

```
src/Data/Singletons/Prelude/Base.hs:54:3: error:
    • Couldn't match expected type ‘p0’ with actual type ‘Sing t1’
        ‘p0’ is untouchable
          inside the constraints: z0 ~ '[]
          bound by a pattern with constructor: SNil :: forall k. Sing '[],
                   in an equation for ‘sGo’
          at src/Data/Singletons/Prelude/Base.hs:(54,3)-(99,3)
    • In the expression: sZ
      In an equation for ‘sGo’: sGo SNil = sZ
      In the first argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘(let
            sGo SNil = sZ
            sGo (SCons (sY :: Sing y_a1q7a) (sYs :: Sing ys_a1q7b))
              = (Data.Singletons.Internal.applySing
                   ((Data.Singletons.Internal.applySing sK) sY))
                  ((Data.Singletons.Internal.applySing
                      ((Data.Singletons.Internal.singFun1
                          @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
                         sGo))
                     sYs)
          in
            (Data.Singletons.Internal.singFun1
               @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
              sGo)’
    • Relevant bindings include
        sGo :: Sing z0 -> p0
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sZ :: Sing t1 (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sK :: Sing t (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sFoldr :: Sing t
                  -> Sing t1
                  -> Sing t2
                  -> Sing
                       (Data.Singletons.Internal.Apply
                          (Data.Singletons.Internal.Apply
                             (Data.Singletons.Internal.Apply FoldrSym0 t) t1)
                          t2)
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
   |
54 | $(singletonsOnly [d|
   |   ^^^^^^^^^^^^^^^^^^...

src/Data/Singletons/Prelude/Base.hs:54:3: error:
    • Couldn't match type ‘p0’
                     with ‘Sing
                             (Data.Singletons.Internal.Apply
                                (Data.Singletons.Internal.Apply t n)
                                (Let6989586621679348488Go t t1 t2 n1))’
        because type variables ‘n’, ‘n1’ would escape their scope
      These (rigid, skolem) type variables are bound by
        a pattern with constructor:
          SCons :: forall a (n1 :: a) (n2 :: [a]).
                   Sing n1 -> Sing n2 -> Sing (n1 : n2),
        in an equation for ‘sGo’
        at src/Data/Singletons/Prelude/Base.hs:(54,3)-(99,3)
      Expected type: p0
        Actual type: Sing
                       (Data.Singletons.Internal.Apply t n
                        Data.Singletons.Internal.@@ Let6989586621679348488Go t t1 t2 n1)
    • In the expression:
        (Data.Singletons.Internal.applySing
           ((Data.Singletons.Internal.applySing sK) sY))
          ((Data.Singletons.Internal.applySing
              ((Data.Singletons.Internal.singFun1
                  @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
                 sGo))
             sYs)
      In an equation for ‘sGo’:
          sGo (SCons (sY :: Sing y_a1q7a) (sYs :: Sing ys_a1q7b))
            = (Data.Singletons.Internal.applySing
                 ((Data.Singletons.Internal.applySing sK) sY))
                ((Data.Singletons.Internal.applySing
                    ((Data.Singletons.Internal.singFun1
                        @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
                       sGo))
                   sYs)
      In the first argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘(let
            sGo SNil = sZ
            sGo (SCons (sY :: Sing y_a1q7a) (sYs :: Sing ys_a1q7b))
              = (Data.Singletons.Internal.applySing
                   ((Data.Singletons.Internal.applySing sK) sY))
                  ((Data.Singletons.Internal.applySing
                      ((Data.Singletons.Internal.singFun1
                          @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
                         sGo))
                     sYs)
          in
            (Data.Singletons.Internal.singFun1
               @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
              sGo)’
    • Relevant bindings include
        sYs :: Sing n1 (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sY :: Sing n (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sGo :: Sing z0 -> p0
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sA_6989586621679348483 :: Sing t2
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sZ :: Sing t1 (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sK :: Sing t (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        (Some bindings suppressed; use -fmax-relevant-binds=N or -fno-max-relevant-binds)
   |
54 | $(singletonsOnly [d|
   |   ^^^^^^^^^^^^^^^^^^...

src/Data/Singletons/Prelude/Base.hs:54:3: error:
    • Could not deduce: t3 ~ (n : n1)
      from the context: z0 ~ (n : n1)
        bound by a pattern with constructor:
                   SCons :: forall a (n1 :: a) (n2 :: [a]).
                            Sing n1 -> Sing n2 -> Sing (n1 : n2),
                 in an equation for ‘sGo’
        at src/Data/Singletons/Prelude/Base.hs:(54,3)-(99,3)
      ‘t3’ is a rigid type variable bound by
        a type expected by the context:
          Data.Singletons.Internal.SingFunction1
            (Let6989586621679348488GoSym3 t t1 t2)
        at src/Data/Singletons/Prelude/Base.hs:(54,3)-(99,3)
      Expected type: Sing t3
                     -> Sing
                          (Let6989586621679348488GoSym3 t t1 t2
                           Data.Singletons.Internal.@@ t3)
        Actual type: Sing z0 -> p0
    • In the second argument of ‘Data.Singletons.Internal.singFun1’, namely
        ‘sGo’
      In the first argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘((Data.Singletons.Internal.singFun1
             @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
            sGo)’
      In the second argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘((Data.Singletons.Internal.applySing
             ((Data.Singletons.Internal.singFun1
                 @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
                sGo))
            sYs)’
    • Relevant bindings include
        sYs :: Sing n1 (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sY :: Sing n (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sGo :: Sing z0 -> p0
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sA_6989586621679348483 :: Sing t2
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sK :: Sing t (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sFoldr :: Sing t
                  -> Sing t1
                  -> Sing t2
                  -> Sing
                       (Data.Singletons.Internal.Apply
                          (Data.Singletons.Internal.Apply
                             (Data.Singletons.Internal.Apply FoldrSym0 t) t1)
                          t2)
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
   |
54 | $(singletonsOnly [d|
   |   ^^^^^^^^^^^^^^^^^^...

src/Data/Singletons/Prelude/Base.hs:54:3: error:
    • Couldn't match type ‘p0’
                     with ‘Sing (Let6989586621679348488Go t t1 t2 t3)’
        because type variable ‘t3’ would escape its scope
      This (rigid, skolem) type variable is bound by
        a type expected by the context:
          Data.Singletons.Internal.SingFunction1
            (Let6989586621679348488GoSym3 t t1 t2)
        at src/Data/Singletons/Prelude/Base.hs:(54,3)-(99,3)
      Expected type: Sing t3
                     -> Sing
                          (Let6989586621679348488GoSym3 t t1 t2
                           Data.Singletons.Internal.@@ t3)
        Actual type: Sing z0 -> p0
    • In the second argument of ‘Data.Singletons.Internal.singFun1’, namely
        ‘sGo’
      In the expression:
        (Data.Singletons.Internal.singFun1
           @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
          sGo
      In the first argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘(let
            sGo SNil = sZ
            sGo (SCons (sY :: Sing y_a1q7a) (sYs :: Sing ys_a1q7b))
              = (Data.Singletons.Internal.applySing
                   ((Data.Singletons.Internal.applySing sK) sY))
                  ((Data.Singletons.Internal.applySing
                      ((Data.Singletons.Internal.singFun1
                          @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
                         sGo))
                     sYs)
          in
            (Data.Singletons.Internal.singFun1
               @(Let6989586621679348488GoSym3 k_a1q6N z_a1q6O a_6989586621679348483_a1q6P))
              sGo)’
    • Relevant bindings include
        sGo :: Sing z0 -> p0
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sA_6989586621679348483 :: Sing t2
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sZ :: Sing t1 (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sK :: Sing t (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
        sFoldr :: Sing t
                  -> Sing t1
                  -> Sing t2
                  -> Sing
                       (Data.Singletons.Internal.Apply
                          (Data.Singletons.Internal.Apply
                             (Data.Singletons.Internal.Apply FoldrSym0 t) t1)
                          t2)
          (bound at src/Data/Singletons/Prelude/Base.hs:54:3)
   |
54 | $(singletonsOnly [d|
   |   ^^^^^^^^^^^^^^^^^^...
```

Perhaps this mysterious `k_aiSK` variable in the type signature of `sGo` is to blame? I'm not even sure where that's coming from...

**Comment by goldfirere** [comment]
Ah, yes. That is unfortunate. The problem is that GHC (sensibly) can't do type inference in the result type of a GADT pattern match. The original `go` didn't have a GADT pattern-match, and so type inference succeeded. But that doesn't work in our output.

Would a partial type signature work?

```
sGo :: _ -> Sing (Apply (GoSym3 _ _ _) _)
```

?

I'm OK with requiring users to enable `-XPartialTypeSignatures` if that's really the only way forward.

**Comment by RyanGlScott** [comment]
I don't think the `PartialTypeSignatures` approach would even fix the problem. For starters, if I use your suggested replacement for `sGo`'s type signature:

```haskell
sFoldr ::
  forall (t_a6nE :: TyFun a_a6mQ (TyFun b_a6mR b_a6mR
                                  -> Type)
                    -> Type)
         (t_a6nF :: b_a6mR)
         (t_a6nG :: [a_a6mQ]).
  Sing t_a6nE
  -> Sing t_a6nF
     -> Sing t_a6nG
        -> Sing (Apply (Apply (Apply FoldrSym0 t_a6nE) t_a6nF) t_a6nG :: b_a6mR)
sFoldr
  (sK :: Sing k_a6nf)
  (sZ :: Sing z_a6ng)
  (sA_6989586621679034295 :: Sing a_6989586621679034295_a6nh)
  = (applySing
       (let
          sGo :: _ -> Sing (Apply (Let6989586621679034300GoSym3 _ _ _) _)
          sGo SNil = sZ
          sGo
            (SCons (sY :: Sing y_a6nC)
                                                                      (sYs :: Sing ys_a6nD))
            = (applySing ((applySing sK) sY))
                ((applySing
                    ((singFun1
                        @(Let6989586621679034300GoSym3 k_a6nf z_a6ng a_6989586621679034295_a6nh))
                       sGo))
                   sYs)
        in
          (singFun1
             @(Let6989586621679034300GoSym3 k_a6nf z_a6ng a_6989586621679034295_a6nh))
            sGo))
      sA_6989586621679034295
```

Then it'll fail in a different way:

```
Bug.hs:99:11: error:
    • Couldn't match type ‘Let6989586621679034300Go w0 w1 w2 w3’
                     with ‘Let6989586621679034300Go w w4 w5 w6’
      Expected type: Sing z -> Sing (Let6989586621679034300Go w w4 w5 w6)
        Actual type: Sing z
                     -> Sing (Let6989586621679034300Go w0 w1 w2 w3)
      NB: ‘Let6989586621679034300Go’ is a type function, and may not be injective
      The type variables ‘w0’, ‘w1’, ‘w2’, ‘w3’ are ambiguous
    • In the ambiguity check for the inferred type for ‘sGo’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      When checking the inferred type
        sGo :: forall k1 k2 a k3 (z :: [a]) (w1 :: k1
                                                   ~> (k2
                                                       ~> k2)) (w2 :: k2) (w3 :: k3) (w4 :: [k1]).
               Sing z -> Sing (Let6989586621679034300Go w1 w2 w3 w4)
      In the first argument of ‘applySing’, namely
        ‘(let
            sGo :: _ -> Sing (Apply (Let6989586621679034300GoSym3 _ _ _) _)
            sGo SNil = sZ
            sGo (SCons (sY :: Sing y_a6nC) (sYs :: Sing ys_a6nD))
              = (applySing ((applySing sK) sY))
                  ((applySing
                      ((singFun1
                          @(Let6989586621679034300GoSym3 k_a6nf z_a6ng a_6989586621679034295_a6nh))
                         sGo))
                     sYs)
          in
            (singFun1
               @(Let6989586621679034300GoSym3 k_a6nf z_a6ng a_6989586621679034295_a6nh))
              sGo)’
   |
99 |           sGo SNil = sZ
   |           ^^^^^^^^^^^^^...
```

Moreover, this wouldn't even fix the [original program](https://github.com/goldfirere/singletons/issues/296#issue-292531016), would it? After all, the problematic type signature was:

```
sX :: Sing Let6989586621679034054XSym0
```

But this has no arguments, so this wouldn't even be changed under the new scheme, leaving us in the same situation as before.

**Comment by RyanGlScott** [comment]
Given that the approach in https://github.com/goldfirere/singletons/issues/296#issuecomment-363466900 is unlikely to work, I'm still inclined to try my suggestion at the bottom of https://github.com/goldfirere/singletons/issues/296#issue-292531016, which I'll spell out more clearly. Namely, we should:

* Keep track of the bound kind variables during promotion
* When promoting a `let`-expression, put all of the bound variables in scope as arguments to the `Let` type family. For instance, we have in the original example:

   ```haskell
    type family Let6989586621679034055Z :: MyProxy_a6jb a_a6j9 where
      Let6989586621679034055Z = MyProxySym0
   ```

   With this proposed change, this would become:

   ```haskell
    type family Let6989586621679034055Z a_a6j9 :: MyProxy_a6jb a_a6j9 where
      Let6989586621679034055Z = MyProxySym0
   ```

   And propagate similar changes when giving singled expressions type signatures involving promoted `let`s.

This would nip the ambiguity in the bud and wouldn't require `PartialTypeSignatures`. It would make these type families more kind-variable–dense, but I think that would be an acceptable price to pay.

That being said, I'm now forced to [eat my words](https://github.com/goldfirere/singletons/pull/298#discussion_r166021341), since I previously argued that keeping track of bound kind variables during promotion would be too unwieldy. But given that we would need a similar trick to fix this bug, and that this trick must be performed during promotion, I think I'll just have to bite the bullet and do it anyway.

**Comment by RyanGlScott** [comment]
The approach in https://github.com/goldfirere/singletons/issues/296#issuecomment-363823093 is quite promising. I implemented this locally, and almost everything works.

I discovered one surprising gotcha during this process. Imagine you're promoting `foldl`. Before, this would promote to something like:

```haskell
type family Foldl (f :: b ~> a ~> b) (z :: b) (xs :: [a]) :: b where
  Foldl f z0 xs0 = Apply (Apply (LetLgoSym3 f z xs) z) xs
```

Under the new plan, we'd augment `LetLgoSym3` with all of the kinds of `Foldl`, so we'd now need to promote `foldl` as:

```haskell
type family Foldl (f :: b ~> a ~> b) (z :: b) (xs :: [a]) :: b where
  Foldl f z0 xs0 = Apply (Apply (LetLgoSym5 a b f z xs) z) xs
```

But this won't work by itself—the `a` and `b` on the RHS are not bound on the LHS! We end up needing to decorate the LHS types with the kinds from the type family declaration itself:

```haskell
type family Foldl (f :: b ~> a ~> b) (z :: b) (xs :: [a]) :: b where
  Foldl (f :: b ~> a ~> b) (z0 :: b) (xs0 :: [a]) = Apply (Apply (LetLgoSym5 a b f z xs) z) xs
```

And now `Foldl` goes through without issues.

-----

I mentioned that _almost_ everything works—what's the catch? Well, let's look at `LetLgo`:

```haskell
type family LetLgo a b f z0 xs0 (z :: b) (xs :: [a]) :: b where ...
```

The implementation of `LetLgo` is unimportant. What's notable here is that we now explicitly bind the kinds of `z` and `xs` using the power of `TypeInType`. However, we also need to defunctionalize `LetLgo`, but this topples over due to #216, as we generate incorrectly scoped defunctionalization symbols, like:

```haskell
type LetLgoSym7 t_aTkY t_aTkZ t_aTl0 t_aTl1 t_aTl2 (t_aTl3 :: b) (t_aTl4 :: [a]) =
        LetLgo t_aTkY t_aTkZ t_aTl0 t_aTl1 t_aTl2 t_aTl3 t_aTl4
```

So this is blocked on #216.

That being said, I don't think this was a complete waste of time, since I did manage to implement a fix for #297 in the process that tracks bound kind variables during promotion instead of singling. I'll clean up my branch in the coming days and open a proper pull request (which should supersede #298).

**Comment by RyanGlScott** [comment]
Hm... my plan in https://github.com/goldfirere/singletons/issues/296#issuecomment-364629893 might not be fully baked. One thing that that plan requires is annotating LHS type variables with their kinds, such as in this example:

```haskell
type family Foldl (f :: b ~> a ~> b) (z :: b) (xs :: [a]) :: b where
  Foldl (f :: b ~> a ~> b) (z0 :: b) (xs0 :: [a]) = Apply (Apply (LetLgoSym5 a b f z xs) z) xs
```

However, it doesn't appear that this will always be a valid solution. For example, I just found out that terms and types differ in their treatment of pattern signatures. At the term level, it appears that it's _always_ safe to copy the type of a binder into its pattern, such as in here:

```haskell
data T :: Type -> Type where
  MkT :: Bool -> T Bool

foo :: T a -> a
foo (MkT x :: T a) = x
```

Here, it's perfectly safe to copy over `T a` into the pattern signature of `MkT x`. However, if we try the same in a type family:

```haskell
type family Foo (x :: T a) :: a where
  Foo (MkT x :: T a) = x
```

Then GHC balks:

```
    • Expected kind ‘T a’, but ‘MkT x’ has kind ‘T Bool’
    • In the first argument of ‘Foo’, namely ‘(MkT x :: T a)’
      In the type family declaration for ‘Foo’
```

So I'm leery about putting type pattern signatures on everything if it means we might bump into this restriction.

**Comment by RyanGlScott** [comment]
I just realized that there is a much, much simpler way to make this work that doesn't require lambda-lifting kind variables (which would be nice to do on its own, but is blocked on #433). Recall the code that gets generated for `f`:

```hs
type LetZ :: MyProxy a
type family LetZ where
  LetZ = 'MyProxy

type family LetX where
  LetX = LetZ

type F :: forall a. MyProxy a -> MyProxy a
type family F x where
  F 'MyProxy = LetX

sF :: forall a (t :: MyProxy a). Sing t -> Sing (F t :: MyProxy a)
sF SMyProxy =
  let sX :: Sing LetX
      sX = let sZ :: Sing (LetZ :: MyProxy a)
               sZ = SMyProxy in sZ
  in sX
```
```
Bug.hs:48:33: error:
    • Could not deduce: a1 ~ a
      from the context: t ~ 'MyProxy
        bound by a pattern with constructor:
                   SMyProxy :: forall a. SMyProxy 'MyProxy,
                 in an equation for ‘sF’
        at Bug.hs:45:4-11
      ‘a1’ is a rigid type variable bound by
        the type signature for:
          sX :: forall a1. Sing LetX
        at Bug.hs:46:7-21
      ‘a’ is a rigid type variable bound by
        the type signature for:
          sF :: forall a (t :: MyProxy a). Sing t -> Sing (F t)
        at Bug.hs:44:1-66
      Expected type: Sing LetX
        Actual type: Sing LetZ
    • In the expression: sZ
      In the expression:
        let
          sZ :: Sing (LetZ :: MyProxy a)
          sZ = SMyProxy
        in sZ
      In an equation for ‘sX’:
          sX
            = let
                sZ :: Sing (LetZ :: MyProxy a)
                sZ = SMyProxy
              in sZ
    • Relevant bindings include
        sZ :: Sing LetZ (bound at Bug.hs:48:16)
        sX :: Sing LetX (bound at Bug.hs:47:7)
        sF :: Sing t -> Sing (F t) (bound at Bug.hs:45:1)
   |
48 |                sZ = SMyProxy in sZ
   |                                 ^^
```

Why is the error message complaining about `a1`? It's because the type of `sX` is being king-generalized! It's as if we had written this:

```hs
sF :: forall a (t :: MyProxy a). Sing t -> Sing (F t :: MyProxy a)
sF SMyProxy =
  let sX :: forall a1. Sing @a1 LetX
      sX = ...
```

This is most certainly not what we want, especially since the type of `x` does _not_ get generalized in the original definition of `f`. Unfortunately, the type of `sX` does get kind-generalized since we give it an explicit type signature of `Sing LetX`.

Luckily, we have a tool at our disposal for disabling kind generalization: visible kind application. Namely, we can visibly apply a wildcard type in the type of `sX`:

```hs
sF :: forall a (t :: MyProxy a). Sing t -> Sing (F t :: MyProxy a)
sF SMyProxy =
  let sX :: Sing @_ LetX
      sX = ...
```

That is enough to turn off kind generalization and make `sF` typecheck. Hooray!

Now here comes the tricky part: how do we teach `singletons` to generate code with `@_` in the right places? Since `x` lacked a type signature in the original definition of `f`, it's tempting to think that one can insert `Sing @_` when singling any function that lacks a type signature. Unfortunately, this will not always work. For instance, this implementation of `foldr`:

```hs
    foldr                   :: (a -> b -> b) -> b -> [a] -> b
    foldr k z = go
              where
                go []     = z
                go (y:ys) = y `k` go ys
```

If the `where`-bound function `go` is singled such that `sGo`'s type signature contains `Sing @_`, then GHC's type inference is no longer able to handle the body of `sGo`:

<details>

```
src/Data/Singletons/Prelude/Base.hs:56:3: error:
    • Can't quantify over ‘arg_a19jP’
        bound by the partial type signature:
          sGo :: forall arg_a19jP.
                 Sing arg_a19jP
                 -> Sing @_ (Data.Singletons.Internal.Apply (Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he) arg_a19jP)
    • In the first argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘(let
            sGo ::
              forall arg_a19jP.
              Sing arg_a19jP
              -> Sing @_ (Data.Singletons.Internal.Apply (Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he) arg_a19jP)
            sGo (sX :: Sing x_a19hi)
              = let ...
                in
                  (id
                     @(Sing (Case_6989586621679283791 x_a19hi p_a19hc f_a19hd a_6989586621679283773_a19he (Let6989586621679283789Scrutinee_6989586621679283709Sym4 x_a19hi p_a19hc f_a19hd a_6989586621679283773_a19he))))
                    (case sScrutinee_6989586621679283709 of
                       STrue -> sX
                       SFalse
                         -> (Data.Singletons.Internal.applySing
                               ((Data.Singletons.Internal.singFun1
                                   @(Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he))
                                  sGo))
                              ((Data.Singletons.Internal.applySing sF) sX))
          in
            (Data.Singletons.Internal.singFun1
               @(Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he))
              sGo)’
      In the expression:
        (Data.Singletons.Internal.applySing
           (let
              sGo ::
                forall arg_a19jP.
                Sing arg_a19jP
                -> Sing @_ (Data.Singletons.Internal.Apply (Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he) arg_a19jP)
              sGo (sX :: Sing x_a19hi)
                = let ...
                  in
                    (id
                       @(Sing (Case_6989586621679283791 x_a19hi p_a19hc f_a19hd a_6989586621679283773_a19he (Let6989586621679283789Scrutinee_6989586621679283709Sym4 x_a19hi p_a19hc f_a19hd a_6989586621679283773_a19he))))
                      (case sScrutinee_6989586621679283709 of
                         STrue -> sX
                         SFalse
                           -> (Data.Singletons.Internal.applySing
                                 ((Data.Singletons.Internal.singFun1
                                     @(Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he))
                                    sGo))
                                ((Data.Singletons.Internal.applySing sF) sX))
            in
              (Data.Singletons.Internal.singFun1
                 @(Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he))
                sGo))
          sA_6989586621679283773
      In an equation for ‘sUntil’:
          sUntil
            (sP :: Sing p_a19hc)
            (sF :: Sing f_a19hd)
            (sA_6989586621679283773 :: Sing a_6989586621679283773_a19he)
            = (Data.Singletons.Internal.applySing
                 (let
                    sGo ::
                      forall arg_a19jP.
                      Sing arg_a19jP
                      -> Sing @_ (Data.Singletons.Internal.Apply (Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he) arg_a19jP)
                    sGo (sX :: Sing x_a19hi)
                      = let ...
                        in
                          (id
                             @(Sing (Case_6989586621679283791 x_a19hi p_a19hc f_a19hd a_6989586621679283773_a19he (Let6989586621679283789Scrutinee_6989586621679283709Sym4 x_a19hi p_a19hc f_a19hd a_6989586621679283773_a19he))))
                            (case sScrutinee_6989586621679283709 of
                               STrue -> ...
                               SFalse -> ...)
                  in
                    (Data.Singletons.Internal.singFun1
                       @(Let6989586621679283785GoSym3 p_a19hc f_a19hd a_6989586621679283773_a19he))
                      sGo))
                sA_6989586621679283773
   |
56 | $(singletonsOnly [d|
   |   ^^^^^^^^^^^^^^^^^^...

src/Data/Singletons/Prelude/Base.hs:56:3: error:
    • Could not deduce: t3 ~ (n : n1)
      from the context: arg ~ (n : n1)
        bound by a pattern with constructor:
                   SCons :: forall a (n1 :: a) (n2 :: [a]).
                            Sing n1 -> Sing n2 -> SList (n1 : n2),
                 in an equation for ‘sGo’
        at src/Data/Singletons/Prelude/Base.hs:(56,3)-(111,3)
      ‘t3’ is a rigid type variable bound by
        a type expected by the context:
          Data.Singletons.Internal.SingFunction1
            (Let6989586621679283886GoSym3 t t1 t2)
        at src/Data/Singletons/Prelude/Base.hs:(56,3)-(111,3)
      Expected type: Sing t3
                     -> Sing
                          (Let6989586621679283886GoSym3 t t1 t2
                           Data.Singletons.Internal.@@ t3)
        Actual type: Sing arg
                     -> Sing
                          (Data.Singletons.Internal.Apply
                             (Let6989586621679283886GoSym3 t t1 t2) arg)
    • In the second argument of ‘Data.Singletons.Internal.singFun1’, namely
        ‘sGo’
      In the first argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘((Data.Singletons.Internal.singFun1
             @(Let6989586621679283886GoSym3 k_a19iP z_a19iQ a_6989586621679283874_a19iR))
            sGo)’
      In the second argument of ‘Data.Singletons.Internal.applySing’, namely
        ‘((Data.Singletons.Internal.applySing
             ((Data.Singletons.Internal.singFun1
                 @(Let6989586621679283886GoSym3 k_a19iP z_a19iQ a_6989586621679283874_a19iR))
                sGo))
            sYs)’
    • Relevant bindings include
        sYs :: Sing n1 (bound at src/Data/Singletons/Prelude/Base.hs:56:3)
        sY :: Sing n (bound at src/Data/Singletons/Prelude/Base.hs:56:3)
        sGo :: Sing arg
               -> Sing
                    (Data.Singletons.Internal.Apply
                       (Let6989586621679283886GoSym3 t t1 t2) arg)
          (bound at src/Data/Singletons/Prelude/Base.hs:56:3)
        sA_6989586621679283874 :: Sing t2
          (bound at src/Data/Singletons/Prelude/Base.hs:56:3)
        sK :: Sing t (bound at src/Data/Singletons/Prelude/Base.hs:56:3)
        sFoldr :: Sing t
                  -> Sing t1
                  -> Sing t2
                  -> Sing
                       (Data.Singletons.Internal.Apply
                          (Data.Singletons.Internal.Apply
                             (Data.Singletons.Internal.Apply FoldrSym0 t) t1)
                          t2)
          (bound at src/Data/Singletons/Prelude/Base.hs:56:3)
   |
56 | $(singletonsOnly [d|
   |   ^^^^^^^^^^^^^^^^^^...
```
</details>

Blast. GADT pattern matches strike again. But we ccan compromise and only put `_@` on type signatures for functions that don't have any arguments. That way, we wouldn't run the risk of GADT pattern matches affecting type inference—you can't match on a GADT argument if there are no arguments to begin with!

**Comment by goldfirere** [comment]
I've approved the PR, but this is gross and wouldn't stand in a production compiler. It's gross in that it uses a heuristic, making 0 arguments a special case. (Of course, using `@_` just to suppress kind generalization is also gross, but that's a forced grossness based on our very inadequate target language of GHC/Haskell.) I don't have a better idea, and the Note in the PR describes this accurately.

**Comment by RyanGlScott** [comment]
I agree that this is gross, but it does fix a very real problem (namely, the inability to single common uses of as-patterns). I don't have a better idea either, so I'll just say that this was fixed in #440.
